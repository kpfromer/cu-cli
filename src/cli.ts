import { CUSession, ICourse, ITerm } from 'cu-api';
import * as Conf from 'conf';
import moment from 'moment';
import { authorize } from './calendar';
import { google } from 'googleapis';
import { toDays } from './helper';

function getLogin(config: Conf): { username: string; password: string } {
  if (!config.has('username') || !config.has('password'))
    throw new TypeError('User is not loaded in to CU.');
  return {
    username: config.get('username'),
    password: config.get('password')
  };
}

/**
 * Creates a CU session with username and password from system keyring.
 */
async function createSession(config: Conf): Promise<CUSession> {
  const { username, password } = getLogin(config);
  const session = new CUSession();
  await session.init(username, password);
  return session;
}

/**
 * Returns the gpa as a number.
 * @param config the config for the application.
 */
export async function getGpa(config: Conf): Promise<number> {
  const { username, password } = await getLogin(config);
  const session = new CUSession();
  await session.init(username, password);
  const gpa = parseFloat((await session.GPA()).cum_GPA);
  return gpa;
}

export async function getCourses(
  config: Conf,
  term: 'current' | 'next' | 'next-next' | 'previous' = 'current'
): Promise<ICourse[]> {
  const session = await createSession(config);
  const terms = await session.termData();
  let wantedTerm: ITerm['attributeName'] = 'CURRENT_TERM';

  if (term === 'next') wantedTerm = 'NEXT_TERM';
  else if (term === 'next-next') wantedTerm = 'NEXT_NEXT_TERM';
  else if (term === 'previous') wantedTerm = 'PREVIOUS_TERM';

  const foundTerm = terms.find((term) => term.attributeName === wantedTerm);
  if (!foundTerm) {
    throw new Error('Could not find current term data.');
  }
  const courses = Array.from(
    (await session.classTermData(foundTerm.term4)).values()
  );
  return courses;
}

/**
 * Syncs classes from CU term to google calendar.
 */
export async function syncClassesCalendar(
  config: Conf,
  term: 'current' | 'next' | 'next-next' | 'previous' = 'current'
): Promise<void> {
  const courses = await getCourses(config, term);

  if (!config.has('google')) throw new Error('Not logged in to google.');
  const auth = await authorize(config.get('google'));
  const calendar = google.calendar({ version: 'v3', auth });

  for (let course of courses) {
    const startTime = moment(
      `${course.courseStartDate} ${course.courseStartTime}`,
      'YYYY-MM-DD k:m'
    );
    const endTime = moment(
      `${course.courseStartDate} ${course.courseStopTime}`,
      'YYYY-MM-DD k:m'
    );
    const endDate = moment(course.courseStopDate, 'YYYY-MM-DD');
    // const start = course.courseStartDate
    const days = toDays(course.days);
    const instructorText = course.instructors.reduce(
      (prev, instructor, index) => {
        if (index === 0) {
          return `\t${instructor.name} - ${instructor.name}`;
        }
        return prev + `\n\t${instructor.name} - ${instructor.email}`;
      },
      ''
    );
    console.log(`Creating "${course.courseTitle}" event.`);
    await calendar.events.insert({
      calendarId: 'primary',
      requestBody: {
        summary: course.courseTitle,
        description: `Course Id: ${course.courseId}\nSubject: ${course.courseSubject}${course.courseNumber}-${course.courseSection}\nCredits: ${course.credits}\nInstructors:\n${instructorText}`,
        start: {
          dateTime: startTime.toISOString(), // yyyy-mm-dd
          timeZone: 'America/Denver'
        },
        end: {
          dateTime: endTime.toISOString(), // yyyy-mm-dd
          timeZone: 'America/Denver'
        },
        recurrence: [
          `RRULE:FREQ=WEEKLY;UNTIL=${endDate
            .locale('en')
            .format('YYYYMMDD[T]000000[Z]')};WKST=SU;BYDAY=${days.join(',')}`
        ],
        location: course.descrLocation
      }
    });
  }
}
