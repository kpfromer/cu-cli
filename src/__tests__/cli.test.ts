jest.mock('cu-api');
import * as CU from 'cu-api';

import { getGpa, getCourses } from '../cli';

describe('getGpa', () => {
  let config: any;
  beforeEach(() => {
    jest.clearAllMocks();
    config = {
      get: jest.fn(),
      has: jest.fn()
    };
    config.get.mockImplementation((name: string) => {
      if (name === 'username') return 'mock-username';
      else if (name === 'password') return 'mock-password';
      else throw new Error();
    });
    config.has.mockImplementation((name: string) => {
      return name === 'username' || name === 'password';
    });
  });
  it('returns the gpa', async () => {
    expect.assertions(3);
    (CU as jest.Mocked<
      typeof CU
    >).CUSession.prototype.GPA.mockResolvedValueOnce({
      cum_GPA: '3.64',
      cur_GPA: '4.0'
    });

    const gpa = await getGpa(config);

    const CUSessionInstance = (CU as jest.Mocked<typeof CU>).CUSession.mock
      .instances[0];
    expect(CU.CUSession).toHaveBeenCalled();
    expect(CUSessionInstance.init).toHaveBeenCalledWith(
      'mock-username',
      'mock-password'
    );
    expect(gpa).toEqual(3.64);
  });
});

describe('getCourses', () => {
  let config: any;
  let termData: CU.ITerm[];
  beforeEach(() => {
    jest.clearAllMocks();
    config = {
      get: jest.fn(),
      has: jest.fn()
    };
    config.get.mockImplementation((name: string) => {
      if (name === 'username') return 'mock-username';
      else if (name === 'password') return 'mock-password';
      else throw new Error();
    });
    config.has.mockImplementation((name: string) => {
      return name === 'username' || name === 'password';
    });
    // mock term metadata
    termData = [
      {
        term4: '2200',
        term5: 'term5',
        startDate: '2019-08-01',
        endDate: '2020-12-25',
        termFriendly: 'Fall 2019',
        attributeName: 'PREVIOUS_TERM'
      },
      {
        term4: '2201',
        term5: 'term5',
        startDate: '2020-01-01',
        endDate: '2020-04-04',
        termFriendly: 'Spring 2020',
        attributeName: 'CURRENT_TERM'
      },
      {
        term4: '2202',
        term5: 'term5',
        startDate: '2020-05-01',
        endDate: '2020-08-01',
        termFriendly: 'Summer 2020',
        attributeName: 'NEXT_TERM'
      },
      {
        term4: '2203',
        term5: 'term5',
        startDate: '2020-08-01',
        endDate: '2020-12-25',
        termFriendly: 'Fall 2020',
        attributeName: 'NEXT_NEXT_TERM'
      }
    ];
  });
  it("gets current term's courses", async () => {
    expect.assertions(4);

    const classTermData = new Map();
    classTermData.set('class-id', {
      name: 'Computer Systems'
    });

    (CU as jest.Mocked<
      typeof CU
    >).CUSession.prototype.termData.mockResolvedValueOnce(termData);

    (CU as jest.Mocked<
      typeof CU
    >).CUSession.prototype.classTermData.mockResolvedValueOnce(classTermData);

    const courses = await getCourses(config);

    const CUSessionInstance = (CU as jest.Mocked<typeof CU>).CUSession.mock
      .instances[0];
    expect(CU.CUSession).toHaveBeenCalled();
    expect(CUSessionInstance.init).toHaveBeenCalledWith(
      'mock-username',
      'mock-password'
    );
    expect(CUSessionInstance.classTermData).toHaveBeenCalledWith('2201'); // current term
    expect(courses).toEqual([classTermData.get('class-id')]);
  });
  it('errors if term is not found', () => {
    (CU as jest.Mocked<
      typeof CU
    >).CUSession.prototype.termData.mockResolvedValueOnce([
      {
        term4: '2200',
        term5: 'term5',
        startDate: '2019-08-01',
        endDate: '2020-12-25',
        termFriendly: 'Fall 2019',
        attributeName: 'PREVIOUS_TERM'
      }
    ]);

    const errorWrapper = async () => {
      await getCourses(config);
    };

    expect(errorWrapper).rejects.toThrowErrorMatchingSnapshot();
  });
});
