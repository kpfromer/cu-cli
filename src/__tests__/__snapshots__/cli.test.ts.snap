// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCourses errors if term is not found 1`] = `"Could not find current term data."`;

exports[`syncClassesCalendar creates calendar events based on classes 1`] = `
Object {
  "calendarId": "primary",
  "requestBody": Object {
    "description": "Course Id: courseId
Subject: CSCI2824-666
Credits: 55
Instructors:
	Kyle Pfromer - Kyle Pfromer",
    "end": Object {
      "dateTime": "2020-01-01T16:50:00.000Z",
      "timeZone": "America/Denver",
    },
    "location": "engineering center",
    "recurrence": Array [
      "RRULE:FREQ=WEEKLY;UNTIL=20200404T000000Z;WKST=SU;BYDAY=MO,WE,FR",
    ],
    "start": Object {
      "dateTime": "2020-01-01T16:00:00.000Z",
      "timeZone": "America/Denver",
    },
    "summary": "The is the class name!",
  },
}
`;
